src/main/java/org/opensearch/security/resources/ResourceSharingIndexHandler.java

```diff
@@ -0,0 +1,1338 @@
+        // TODO: Once stashContext is replaced with switchContext this call will have to be modified
+        try (ThreadContext.StoredContext ctx = this.threadPool.getThreadContext().stashContext()) {
+            ResourceSharing entry = new ResourceSharing(resourceIndex, resourceId, createdBy, shareWith);
+
+            IndexRequest ir = client.prepareIndex(resourceSharingIndex)
+                .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+                .setSource(entry.toXContent(jsonBuilder(), ToXContent.EMPTY_PARAMS))
+                .setOpType(DocWriteRequest.OpType.CREATE) // only create if an entry doesn't exist
+                .request();
+
+            ActionListener<IndexResponse> irListener = ActionListener.wrap(
+                idxResponse -> LOGGER.info(
+                    "Successfully created {} entry for resource {} in index {}.",
+                    resourceSharingIndex,
+                    resourceId,
+                    resourceIndex
+                ),
+                (failResponse) -> {
+                    LOGGER.error(failResponse.getMessage());
+                }
+            );
+            client.index(ir, irListener);
+            return entry;
+        } catch (Exception e) {
+            LOGGER.error("Failed to create {} entry.", resourceSharingIndex, e);
+            throw new OpenSearchStatusException("Failed to create " + resourceSharingIndex + " entry.", RestStatus.INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    /**
+     * Fetches all resource sharing records that match the specified system index. This method retrieves
+     * a get of resource IDs associated with the given system index from the resource sharing index.
+     *
+     * <p>The method executes the following steps:
+     * <ol>
+     *   <li>Creates a search request with term query matching the system index</li>
+     *   <li>Applies source filtering to only fetch resource_id field</li>
+     *   <li>Executes the search with a limit of 10000 documents</li>
+     *   <li>Processes the results to extract resource IDs</li>
+     * </ol>
+     *
+     * <p>Example query structure:
+     * <pre>
+     * {
+     *   "query": {
+     *     "term": {
+     *       "source_idx": "resource_index_name"
+     *     }
+     *   },
+     *   "_source": ["resource_id"],
+     *   "size": 10000
+     * }
+     * </pre>
+     *
+     * @param pluginIndex The source index to match against the source_idx field
+     * @param listener    The listener to be notified when the operation completes.
+     *                    The listener receives a set of resource IDs as a result.
+     * @apiNote This method:
+     * <ul>
+     *   <li>Uses source filtering for optimal performance</li>
+     *   <li>Performs exact matching on the source_idx field</li>
+     *   <li>Returns an empty get instead of throwing exceptions</li>
+     * </ul>
+     */
+    public void fetchAllDocuments(String pluginIndex, ActionListener<Set<String>> listener) {
```

Comment:
I think we should handle this case w/o getting all resource IDs. It looks like this is used in cases where the logged in user has all resources visible to them? (i.e. superuser or cluster admin)

If that's the case then can we communicate with the resource plugin to perform an search unrestricted.

Reply:
we are only returning the ids. This method fetches resource-ids from the individual entries in resource-sharing-index and not the pluginIndex.

i;ve renamed this to make it clearer.